engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReplayStatePropertyTest.java

```diff
@@ -99,6 +99,7 @@ private void stopAndRestartEngineAndCompareStates() {
             () ->
                 assertThat(engineRule.getStreamProcessor(1).getCurrentPhase().join())
                     .isEqualTo(Phase.PROCESSING));
+    waitForProcessingToStop();
```

Comment:
Not sure this is sufficient. The thing I don't know is whether there are other actors adding commands independently. E.g. a delete all outdated messages timer that enqueues additional commands. If so the restarted engine could advance past the point it was stopped at. And then we would compare two different states again,

Reply:
Not sure whether I get your comment sorry. I tried it with the changes and without the await of same positions but there is still the possibility of occuring . With the waiting this should not longer the case, since the state Processing is set after the positions are updated and we wait then until they become equal.

Reply:
Not sure I have a point to begin with.

Let's say we are at position 10. We record the state and stop the engine.
Then we restart again. We should end up in position 10 again.

Where I think a potential race condition could be (and again I am not sure about this), is if there are other sources of commands besides the test itself. In a normal engine there seem to be actors running which periodically delete timed out messages, trigger timer events, fail jobs on timeout. I don't fully understand how it works, but it seems like these actors run independently and interact with the state by issuing commands.

If any of those actors is also running in the test engine, then this actor could issue a new command after we restarted the engine and before "processing stopped" and we compare the state. This command would then take position 11. And since we now wait for processing to stop, we would wait until we have processed position 11. Then, when we compare both states we compare states at different positions.

Maybe my concern is invalid, because in the test setup the test is the only piece of code issuing commands. 

But I have a nagging feeling that we have no full control over the engine (e.g. telling it to process until a certain record and then stop).

I hope I explained it better this time, so you can judge whether there is merit to my concern or not.